table/data_block_hash_index.cc

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2011-present, Facebook, Inc. All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+#include <string>
+#include <vector>
+
+#include "rocksdb/slice.h"
+#include "table/data_block_hash_index.h"
+#include "util/coding.h"
+#include "util/hash.h"
+
+namespace rocksdb {
+
+const uint32_t kSeed = 2018;
+const uint32_t kSeed_tag = 214; /* second hash seed */
+
+inline uint16_t HashToBucket(const Slice& s, uint16_t num_buckets) {
+  return (uint16_t)rocksdb::Hash(s.data(), s.size(), kSeed) % num_buckets;
+}
+
+void DataBlockHashIndexBuilder::Add(const Slice& key,
+                                    const uint16_t& restart_index) {
+  uint16_t idx = HashToBucket(key, num_buckets_);
+  /* push a TAG to avoid false postive */
+  /* the TAG is the hash function value of another seed */
+  uint16_t tag = static_cast<uint16_t>(
+      rocksdb::Hash(key.data(), key.size(), kSeed_tag));
```

Comment:
rocksdb::Hash() returns uint32_t ... so what affect will the loss of precision have on our algorithm?

Reply:
Casting to uint16 (effectively chops the higher 16 bit of hash off) has the benefit of reducing the space overhead as well as a more compact bucket that is more CPU cache line friendly. The downside is that there will be more hash collision. The current experiment shows that reducing the TAG length improves throughput. So I assume the benefit overweights the downside.